Add initial Euphonic-Brille design document (#14)
pace-neutrons/pace-developers: euphonic/design/06_brille_integration.md

and linear interpolation within an irreducible part of the first Brillouin
zone. In the context of Euphonic it means that it can be used to linearly
interpolate phonon eigenvalues/vectors to calculate them more quickly. There
are 2 ways in which this could work. One is where Brille calls Euphonic, which
means it can be used for Horace cuts, and Euphonic doesn't get any added
complexity, but also means Euphonic standalone users don't get any of the
benefits. The other option is that Euphonic calls Brille, this would require
changes to Euphonic but would allow faster interpolation for both Horace and
Euphonic standalone users. However, these options are not necessarily exclusive
and it's possible that both could be implemented.

The purpose of this document is to assess the benefits and drawbacks of these
options and define any desired syntax, which will hopefully reveal any
potential issues. This document should also be considered in the context of
what the Brille-SpinW interface will look like, as it would be beneficial to
Horace users if they were somewhat similar.

### General Note on Syntax

There are two options for the syntax, one is to  try to keep the commands as

> specific as possible so each command has fewer arguments and its clearer what

Reviewer: ```suggestion
specific as possible so each command has fewer arguments and it is clearer what
```

option is to try to have as few calls as possible and try to encapsulate all
the options in arguments to one or two functions. This would be easy for a user
who is happy with default arguments, but could become very complicated as
more arguments and features are added (e.g. powder averaging).

## Calling Euphonic from Brille

### Use Case 1: Simulating Horace cuts
This would involve using Horace's simulation functions to call Brillem (which
calls BrillEu, which calls Brille and Euphonic).

**Example Syntax**

This syntax tries to keep the Brille/Euphonic arguments deliberate and
separate, however it could be confusing for some users (nested cell arrays)
and some settings you may want to apply for both Brille and Euphonic rather
than having to set them separately (e.g. if you want to use 4 threads in
Euphonic to create the lookup table, you probably also want to use 4 threads in
Brille to linearly interpolate the eigenvalues/vectors). In addition, with the
current Horace simulation functions the syntax splits arguments up into 'model
parameters' for fitting and 'extra parameters' for constant arguments. For now
I've made an example with the scaling factor as a fitting parameter.

> ```

Reviewer: ```suggestion
```matlab
```

   but do we really want to maintain 2 interfaces to Horace?
 - In the current implementation of this there is some repetition of code to
   enable Brille to wrap a lot of Euphonic's functions, and get the additional
   symmetry information it needs. I'm not sure if this can be avoided for the 
   'calling Euphonic from Brille' case.

## Calling Brille from Euphonic

### Use case 1: Brille 'Force Constants' for Faster Interpolation

This would involve adding another object to Euphonic, for now I'll call it
`BrilleForceConstants` (although a better name might be needed) which would act
like `ForceConstants`, but would perform linear instead of Fourier
interpolation. To make the most of this the current `ForceConstants` in
Euphonic might have to be updated to enable direct calculation of structure
factors, debye waller etc. from `ForceConstants` (without intermediate
eigenvectors), although this might not be required if calculations are
appropriately chunked. Another update to make this integration easier would be
adding symmetry information to Euphonic's `Crystal` object (so that Brille
doesn't need to find it itself).

**Example Syntax**

From Python:

> ```

Reviewer: ```suggestion
```python
```

from euphonic import ureg, ForceConstants
from euphonic.util import mp_grid

fc = ForceConstants.from_castep('quartz.castep_bin')

# Create BrilleForceConstants object
interpolate_kwargs = {'use_c': True, 'n_threads': 4, 'asr': 'reciprocal'}
bfc = fc.calculate_brille_force_constants(
    grid='mesh', max_points=10000, **interpolate_kwargs)

# Create DebyeWaller
dw_phonons = bfc.calculate_qpoint_phonon_modes(mp_grid([6,6,6]))
dw = dw_phonons.calculate_debye_waller(temperature=300*ureg('K'))

# Calculate StructureFactor
phonons = bfc.calculate_qpoint_phonon_modes(qpts, n_threads=4)
sf = phonons.calculate_structure_factor(dw)
```

From Matlab:

The question here is whether the interface should be 'Horace-like' with bespoke
functions to make it easier to simulate cuts, or whether it should more closely
mirror Euphonic's Python interface. The following example is somewhere
inbetween and assumes a Matlab wrapper has been written (named 'meuphonic').

> ```

Reviewer: ```suggestion
```matlab
```

% Create ForceConstants and BrilleForceConstants using Euphonic-like syntax
fc = meuphonic.ForceConstants.from_castep('quartz.castep_bin');
kwargs = {'use_c', true, 'n_threads', 4};
bfc = fc.calculate_brille_force_constants( ...
    'grid', 'mesh', 'max_points', 10000, 'interpolate_kwargs', kwargs);

fit_params = [1.0]; % temperature, scale
pars = {fit_params,
        'force_constants', bfc,  % accepts fc or bfc
        'interpolate_kwargs', {'threads', 4}, % different depending on fc or bfc
        'debye_waller_grid', [6,6,6],
        'temperature', 300.0,
        'conversion_mat', [1,0,0; 0,1,0; 0,0,-1], % Convert Horace Q -> Euphonic Q
        'chunk', 10000};
sim_cut = disp2sqw_eval(exp_cut, @meuphonic.horace_sqw, pars, fwhh);
```

**Advantages:**
 - Would benefit both Horace and Euphonic users
 - Allows Horace users to skip a step depending on if they want to use brille
   or not

**Disadvantages:**
 - Would require additions to Euphonic's ForceConstants, and quite a few
   changes to Brille's existing repos before this option is feasible
 - The interface could end up being quite different to SpinW (does this

>    matter?)

Reviewer: Just a comment on the current SpinW integration. For using SpinW through Brille, the syntax:
```matlab
tri = sw_model('triAF', 1); 
spinw_pars = {'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
brl_pars = {'max_volume', 1e-5};
brl = brillem.Brille(tri, brl_pars{:});
sim_cut = disp2sqw_eval(exp_cut, @brl.horace_sqw, spinw_pars, fwhm);
```
is currently useable. Brille specific options are passed to the constructor `brillem.Brille` and only SpinW related parameters are passed to `disp2sqw_eval`.

For using SpinW directly from Horace, the syntax is:
```matlab
tri = sw_model('triAF', 1); 
spinw_pars = {'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
sim_cut = disp2sqw_eval(exp_cut, @tri.horace_sqw, spinw_pars, fwhm);
```
E.g. the same as with Brille but with the two Brille related lines removed, and calling the `horace_sqw` method of the SpinW object directly. (`tri = sw_model('triAF', 1);` is just a convenient shortcut to create a simple spin wave model. Usually, users would need several lines to create a model themselves.)

The final use case is for users wanting to using Brille with SpinW without using Horace (calculating the linearly interpolated spin wave spectrum without reference to a Horace workspace). The syntax here is:
```matlab
tri = sw_model('triAF', 1); 
spinw_pars = {'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
brl_pars = {'max_volume', 1e-5};
brl = brillem.Brille(tri, brl_pars{:});
brl.fill(spinw_pars{:})
hkl_path = {[0 0 0] [1 1 1] [0 0 1] [0 0 0] 100}; 
spec = brl.spinwave(hkl_path);
sw_plotspec(sw_egrid(sw_neutron(spec), 'Evect', linspace(0,20,100)));
```
compared with the syntax without Brille:
```matlab
tri = sw_model('triAF', 1); 
spinw_pars = {'hermit', false, 'formfact', true, 'resfun', 'gauss'}; 
hkl_path = {[0 0 0] [1 1 1] [0 0 1] [0 0 0] 100}; 
spec = tri.spinwave(hkl_path, spinw_pars{:});
sw_plotspec(sw_egrid(sw_neutron(spec), 'Evect', linspace(0,20,100)));
```
The main difference is that the SpinW parameters has to be passed to the `fill` call which then passes all its arguments to the underlying calculator (e.g. SpinW). No parameters except the hkl path or grid should be passed to the Brillem `spinwave` method unlike for the pure-SpinW case where the SpinW parameters are passed there.

Reviewer: The current Brille/SpinW syntax is very similar to the proposed Euphonic syntax so I think this is fine. The main difference between `brillem.Brille` (the standalone Brille Matlab interface) and `brillem.Euphonic` (the Brille+Euphonic Matlab interface) is that `brillem.Euphonic` is just a light wrapper for `brilleu` (a Python module interfacing Brille to Euphonic) and so does not have many methods to interact with the Brille object. This is pertinent in the separation of keyword arguments or input parameters meant for Brille or for the calculator (Euphonic or SpinW). In the case of `brillem.Euphonic` both sets should be passed to the constructor at the same time; For `brillem.Brille` the constructor does not immediately construct the grid (e.g. invoke the calculator) so it only needs Brille specific arguments. Only secondary calls to `fill` (the `horace_sqw` method implicitly calls `fill` if a grid is not constructed or is out-of-date) need to be passed the arguments for the calculator (SpinW in the above examples). This separation could be useful or annoying depending on your point of view... (it could also make it harder to implement universal arguments which should apply to both Brille and the calculator).

### Use case 2: Brille Q-point Mesh for BZ Integration
This would involve using Brille to produce a mesh of q-points and their weights
which Euphonic could use for Brillouin Zone integration (instead of naively
generating a Monkhorst-Pack grid using `mp_grid` as in previous examples, which
makes no use of symmetry). This is a much 'lighter' use of Brille and should in
theory be easy to implement. However, libraries that Euphonic already depends
on such as `spglib` can already generate weighted q-point Monkhorst-Pack grids,

> so is there actually much to be gained from having a 'smarter' mesh generated

Reviewer 2: ```
so is there actually much to be gained from having a 'smarter' mesh generated
```
@krefson @ajjackson please correct me if I'm wrong, but one of the 'bad' things about MP grids is that they are always constructed from linear combinations of the basis vectors. So if the angle between two basis vectors is unusually small (or large) or the ratio between any two basis vector lengths is large a MP grid may be poorly conditioned for some calculations.
The 'grids' constructed by `brille` are always built in an orthonormal coordinate system as either a Delaunay tetrahedralisation or from a series of (truncated) cubic cells.  So the positions of the points in a `brille` grid (away from the irreducible Brillouin zone boundaries) do not depend on the basis vectors.

Reviewer 3: I think there's a fundamental question to be asked about the merits of linear interpolation for BZ integration. Is it actually going to improve the results? I suppose this is somewhat equivalent to the tetrahedron integration employed in some electronic structure calculations.

The work on generalised grids from Gus Hart's group shows improvement over naive Monkhorst-Pack sampling for electronic structure of realistic materials. https://arxiv.org/pdf/1804.04741.pdf I'd expect this to transfer to phonon sampling as well. I'm not sure how Brille's method compares; it's worth noting that these improved schemes are still based on linear combinations of basis vectors - but use better combinations!

An earlier piece of important work in this area is Moreno/Soler https://journals.aps.org/prb/abstract/10.1103/PhysRevB.45.13891 ; they illustrate the impact of different primitive cell choices. They also recommend optimising the grid origin for maximum symmetry-reduction. This part hasn't been popular; it's no fun to implement and I for one prefer to keep the GP in my calculations because it's usually interesting.
